src/Language/Go/Term.hs

```diff
@@ -0,0 +1,173 @@
+  , Expression.BOr
+  , Expression.BAnd
+  , Expression.BXOr
+  , Expression.LShift
+  , Expression.RShift
+  , Expression.UnsignedRShift
+  , Expression.Complement
+  , Expression.Call
+  , Expression.LessThan
+  , Expression.LessThanEqual
+  , Expression.GreaterThan
+  , Expression.GreaterThanEqual
+  , Expression.Equal
+  , Expression.StrictEqual
+  , Expression.Comparison
+  , Expression.Subscript
+  , Expression.Member
+  , Statement.PostDecrement
+  , Statement.PostIncrement
+  , Expression.MemberAccess
+  , Expression.And
+  , Expression.Not
+  , Expression.Or
+  , Expression.XOr
+  , Go.Syntax.Composite
+  , Go.Syntax.DefaultPattern
+  , Go.Syntax.Defer
+  , Go.Syntax.Field
+  , Go.Syntax.Go
+  , Go.Syntax.Label
+  , Go.Syntax.Package
+  , Go.Syntax.Receive
+  , Go.Syntax.ReceiveOperator
+  , Go.Syntax.Rune
+  , Go.Syntax.Select
+  , Go.Syntax.Send
+  , Go.Syntax.Slice
+  , Go.Syntax.TypeAssertion
+  , Go.Syntax.TypeConversion
+  , Go.Syntax.TypeSwitch
+  , Go.Syntax.TypeSwitchGuard
+  , Go.Syntax.Variadic
+  , Go.Type.BidirectionalChannel
+  , Go.Type.ReceiveChannel
+  , Go.Type.SendChannel
+  , Go.Syntax.Import
+  , Go.Syntax.QualifiedImport
+  , Go.Syntax.SideEffectImport
+  , Literal.Array
+  , Literal.Complex
+  , Literal.Float
+  , Literal.Hash
+  , Literal.Integer
+  , Literal.KeyValue
+  , Literal.Pointer
+  , Literal.Reference
+  , Literal.TextElement
+  , Statement.Assignment
+  , Statement.Break
+  , Statement.Continue
+  , Statement.For
+  , Statement.ForEach
+  , Statement.Goto
+  , Statement.If
+  , Statement.Match
+  , Statement.NoOp
+  , Statement.Pattern
+  , Statement.Return
+  , Statement.Statements
+  , Syntax.Context
+  , Syntax.Error
+  , Syntax.Empty
+  , Syntax.Identifier
+  , Type.Annotation
+  , Type.Array
+  , Type.Function
+  , Type.Interface
+  , Type.Map
+  , Type.Parenthesized
+  , Type.Pointer
+  , Type.Slice
+  , []
+  , Literal.String
+  , Literal.EscapeSequence
+  , Literal.Null
+  , Literal.Boolean
+  ]
+
+
+newtype Term ann = Term { getTerm :: Term.TermF (Sum.Sum Syntax) ann (Term ann) }
+  deriving (Eq, Declarations, FreeVariables, Ord, Show, ToJSON)
+
+instance Term.IsTerm Term where
+  type Syntax Term = Sum.Sum Syntax
+  toTermF = coerce
+  fromTermF = coerce
+
+instance Foldable Term where
+  foldMap = foldMapDefault
+
+instance Functor Term where
+  fmap = fmapDefault
+
+instance Traversable Term where
+  traverse f = go where go = fmap Term . bitraverse f go . getTerm
+
+instance VertexDeclaration Term where
+  toVertex info (Term (Term.In ann syntax)) = toVertex1 ann info syntax
+
+instance Syntax.HasErrors Term where
+  getErrors = cata $ \ (Term.In Loc{..} syntax) ->
+    maybe (fold syntax) (pure . Syntax.unError span) (Sum.project syntax)
+
+
+instance DiffTerms Term where
+  diffTermPair = diffTermPair . bimap (cata Term.Term) (cata Term.Term)
+
+type instance Base (Term ann) = Term.TermF (Sum.Sum Syntax) ann
+
+instance Recursive (Term ann) where
+  project = getTerm
```

Comment:
This is all so beautiful.

Reply:
You do, however, repeat it in a few places. Would `-XDerivingVia` be nice here?

```
newtype Termwise a = Termwise a

instance Traversable a => Functor (Termwise a) where fmap = coerce fmapDefault 
instance Traversable a => Foldable (Termwise a) where foldMap = coerce foldMapDefault
instance IsTerm a => Recursive (Termwise (a ann)) where project = coerce getTerm
```

Then in all these functions you should be able to do a `deriving via`, unless I miss my guess.

Reply:
That should be possible, but again, all of this code is very short-term, so I’m not sure it’s worth the investment.

Reply:
Fair enough, and after the Codex today I’m wary of DerivingVia at all.